### ISSUE TITLE
Material Index not matched

### ISSUE THREAD

**Issue by dongho-shin** [gkjohnson/three-gpu-pathtracer#662]
**Describe the bug**

<img width="325" alt="Untitled (16)" src="https://github.com/user-attachments/assets/9782e35f-c081-44d9-996c-2ba91e90d981">

this model has all the same Materials (8 mesh but share 1 material, share 1 material(same uuid) is a key point of this bug)

1. PathTrace Rendering first after import this model
2. change only material of mesh
<img width="361" alt="image" src="https://github.com/user-attachments/assets/970d95ca-c44c-4e43-8177-b3dc66630d04">

3. PathTrace render again

<img width="406" alt="image" src="https://github.com/user-attachments/assets/4e337267-5ce2-438a-8cb5-c9f9df2942c9">

**Model**

FYI) zip from mac

[Blender_Smeg (2).glb.zip](https://github.com/user-attachments/files/16330712/Blender_Smeg.2.glb.zip)

**Platform:**

 - Device: Desktop
 - OS: MacOS
 - Browser: Chrome
 - Three.js version: r162
 - Library version: 0.0.23

This bug causes here
https://github.com/gkjohnson/three-gpu-pathtracer/blob/d5b1f148afa82d6f9a28be4870e8d9fb664e593c/src/core/utils/GeometryPreparationUtils.js#L46

### sorry for ugly notes
indexOf, if all meshes share 1 materials(in this notes a) geometry.materialIndex attributes are all 0
![Screenshot_20240722_155500_Samsung Notes](https://github.com/user-attachments/assets/f4dbc509-29ff-4463-a218-e8950638bd76)
and then material of mesh has changed(a -> b)
![Screenshot_20240722_155534_Samsung Notes](https://github.com/user-attachments/assets/9aef14e4-f111-49ca-b640-3f91ac74e0b6)

StaticGeometryGenerator, bvhChanged => no changes(because in this scenario no geometry changes) so materialIndex not updated, geometry reuse materialIndex 0 so nothing has been updated

so my sugestion is
before
```js
		const mat = Array.isArray( materials ) ? materials[ group.materialIndex ] : materials;
		const materialIndex = allMaterials.indexOf( mat );
```
after
```js
	//	const mat = Array.isArray( materials ) ? materials[ group.materialIndex ] : materials; // not used
		const materialIndex = group.materialIndex;
```

always welcome any ideas & questions

**Comment by gkjohnson** [comment]
Thanks for reporting and digging in as always!

I suspect this is because the geometry has not changed, causing the static geometry generator to return "NO_CHANGE" meaning the material index buffer attribute is not changed - see [here](https://github.com/gkjohnson/three-gpu-pathtracer/blob/d5b1f148afa82d6f9a28be4870e8d9fb664e593c/src/core/PathTracingSceneGenerator.js#L182-L186).

I think the easy fix would be the following:

```js
// change this
if ( result.changeType !== NO_CHANGE ) {

	updateMaterialIndexAttribute( geometry, materials, materials );

}

// to this
updateMaterialIndexAttribute( geometry, materials, materials );
```

So the material index is always updated. Would you be able to test that?

A more optimal change would be to check if any of the materials have changed before regenerating. But that's a bit more complicated.

**Comment by dongho-shin** [comment]
first try that you suggest me, it fails so I tried workaround weird way
I miss this part
https://github.com/gkjohnson/three-gpu-pathtracer/blob/d5b1f148afa82d6f9a28be4870e8d9fb664e593c/src/core/WebGLPathTracer.js#L353-L365

your suggestion + change of code below fix this issue
```js
//previous
 if ( bvhChanged ) { 
  
 	material.bvh.updateFrom( bvh ); 
 	material.attributesArray.updateFrom( 
 		geometry.attributes.normal, 
 		geometry.attributes.tangent, 
 		geometry.attributes.uv, 
 		geometry.attributes.color, 
 	); 
  
 	material.materialIndexAttribute.updateFrom( geometry.attributes.materialIndex ); 
  
 } 
 //after
  if ( bvhChanged ) { 
  
 	material.bvh.updateFrom( bvh ); 
 	material.attributesArray.updateFrom( 
 		geometry.attributes.normal, 
 		geometry.attributes.tangent, 
 		geometry.attributes.uv, 
 		geometry.attributes.color, 
 	); 
    
 }
 //No matter of bvhChanged
 material.materialIndexAttribute.updateFrom( geometry.attributes.materialIndex );  
```

if you agree this fixed i'll open pr, always thanks for feedback @gkjohnson

**Comment by gkjohnson** [comment]
Good catch - a PR would be great.

It could be improved further if the material list information were saved as a hash to check if the list of materials had actually been updated or not - but I'l leave it up to you if you want to look into that.

**Comment by dongho-shin** [comment]
I'll open a PR this week fixing this issue in this way. if hash improvement of material is not urgent I'll look into it

**Comment by gkjohnson** [comment]
Fixed in #664
